content/blog/2020-03-dplyr-1-0-0-new-verb-features/index.Rmarkdown

```diff
@@ -0,0 +1,231 @@
+
+In this post, I'll introduce you to the major changes in existing dplyr verbs, along with couple of new verbs. I'll give you updates on `summarise()`, `select()` + `rename()`, `mutate()`, and the join functions.
+
+```{r setup, message = FALSE}
+library(dplyr)
+```
+
+## `summarise()` can produce multiple rows and columns
+
+Two big changes make `summarise()` much more flexible. A single summary expression can now return:
+
+* A vector of any length, creating multiple rows.
+* A data frame, creating multiple columns.
+
+In other words, `summarise()` used to be limited to generating a single value (one row, one column), and now we've lifted that restriction so it can generate a rectangle of multiple rows and columns. This change makes `summarise()` as powerful as the now superseded `do()`, and if you were previously solving this type of problem with list-columns allows you to reduce your usage of `tidyr::unnest()`.
+
+This is a big change to `summarise()` but it should have minimal impact on existing code because it _broadens_ the interface: all existing code will continue to work, and a number of inputs that would have previously errored now work. 
+
+### Quantiles
+
+To demonstrate this new feature we'll start by looking at a summary that was previous hard to compute with `summarise()`. `quantile()` was hard to use because it returns multiple values which used to cause `summarise()` to error. Now it's straightforward:
+
+```{r}
+df <- tibble(
+  grp = rep(1:2, each = 10), 
+  x = c(rnorm(10, -0.25, 1), rnorm(10, 0, 1.5)),
+  y = c(rnorm(10, 0.25, 1), rnorm(10, 0, 0.5)),
+)
+df %>% 
+  group_by(grp) %>% 
+  summarise(x = quantile(x, c(0.25, 0.5, 0.75)), q = c(0.25, 0.5, 0.75))
+```
+
+It would be nice to be able to reduce the duplication in this code so that we don't have to type the quantile values twice. We can now write a simple function to do so because `summarise()` expressions can now return multiple columns:
+
+```{r}
+quantile2 <- function(x, q = c(0.25, 0.5, 0.75)) {
+  tibble(x = quantile(x, q), q = q)
+}
+df %>% 
+  group_by(grp) %>% 
+  summarise(quantile2(x, c(0.25, 0.5, 0.75)))
+```
+
+In the past, one of the challenges of writing this sort of function was naming the columns. For example, when you call `quantile2(y)` it'd be nice if you'd get columns `y` and `y_q`, not `x` and `x_q`. Now, thanks to the recent combination of [glue and tidy evaluation](https://www.tidyverse.org/blog/2020/02/glue-strings-and-tidy-eval/) that behaviour is straightforward to implement: 
+
+```{r}
+quantile2 <- function(x, q = c(0.25, 0.5, 0.75)) {
+  tibble("{{ x }}" := quantile(x, q), "{{ x }}_q" := q)
+}
+
+df %>% 
+  group_by(grp) %>% 
+  summarise(quantile2(y, c(0.25, 0.5, 0.75)))
+```
+Figuring out how to name the output columns is a surprisingly complex task and we're still thinking about the best approach. 
```

Comment:
Mention how `set_names()` can be used to override the names if they are not appropriate?